bittensor/_neuron/text/core_validator/__init__.py

```diff
@@ -1362,6 +1373,105 @@ def shapley_base(uids: torch.Tensor, query_responses: List[List[torch.FloatTenso
+                if probs_avg is None:
+                    probs_avg = probs
+                else:
+                    probs_avg += probs
+                probs_k += 1
+
+            except Exception as e:
+                logger.warning(f'Synapse {index_s} error (logits_divergence)\t| '
+                               f'UID {_uid} <dim>[{times[index][index_s]:.2f}s]</dim>: {e}')
+
+    if probs_avg is not None:
+        probs_avg /= probs_k
+        probs_avg_sqrt = probs_avg.sqrt()
+        batch_divergences = []
+
+        # === Distribution divergence ===
+        # Calculate the Hellinger distance (f-divergence) from the average probability distribution for each batch task.
+        for index, _uid in enumerate(uids.tolist()):
+            if return_ops[index][index_s] == bittensor.proto.ReturnCode.Success:
+                try:
+                    probs = topk_tokens_to_vocab_size(query_responses[index][index_s],
+                                                      bittensor.__vocab_size__)  # [batch_size, vocab_size]
+                    divergences = 0.5 * torch.pow(probs.sqrt() - probs_avg_sqrt, 2).sum(dim=1)  # [batch_size] in [0, 1]
+                    divergences = divergences.sqrt()
+                    stats[_uid]['logits_divergences' + ext] = divergences  # [batch_size]
+                    stats[_uid]['logits_divergence' + ext] = divergences.mean()  # scalar
+                    batch_divergences += [divergences]
+
+                except Exception as e:
+                    logger.warning(f'Synapse {index_s} error (logits_divergence)\t| '
+                                   f'UID {_uid} <dim>[{times[index][index_s]:.2f}s]</dim>: {e}')
+
+        batch_divergences = torch.stack(batch_divergences)  # [uids_len, batch_size]
+        avg = batch_divergences.mean(dim=0)  # [batch_size]
+        std = batch_divergences.std(dim=0)  # [batch_size]
+
+        # logger.info(f"Logits divergences: "
+        #             f"avg={', '.join([f'{i}:{v:.3g}' for i, v in enumerate(avg)])}")
+        # logger.info(f"std={', '.join([f'{i}:{v:.3g}' for i, v in enumerate(std)])}")
+
+        # === Calculate divergence excess ===
+        # For each batch task, calculate excess deviation above a single stddev, in terms of stddev,
+        # and apply power to increase score above two stddev, and decrease between one and two stddev.
+        # This will effectively allow zero excess below one stddev, and minimal excess below two stddev,
+        # but amplify any excess above two stddev (only 2.1% of population for normal dist).
+        for _uid, _stats in stats.items():
+            if 'logits_divergences' + ext in _stats:
+                try:
+                    excess = torch.clamp(_stats['logits_divergences' + ext] - (avg + std), 0)  # divergence > avg + std
+                    excess /= std + 1e-9  # stddev multiples above 1 stddev
+                    excess = torch.pow(excess, 3)  # reduce < 2std, increase > 2std
```

Comment:
Would we want this to be parameter on subtensor?

Reply:
Not for this PR, but I think it would be something similar to the synergy/shapley power laws

Reply:
Yes, together with the prune_len, although we'll do this in a future PR.